File: SortVision/public/code/merge/dart/mergeSort.dart
class MergeSort {
  // Time: Best/Avg/Worst O(n log n), Space: O(n)
  static void sort(List<int> arr) {
    if (arr.length <= 1) return;
    _sort(arr, 0, arr.length - 1);
  }

  static void _sort(List<int> arr, int left, int right) {
    if (left >= right) return;
    final mid = left + (right - left) ~/ 2;
    _sort(arr, left, mid);
    _sort(arr, mid + 1, right);
    _merge(arr, left, mid, right);
  }

  static void _merge(List<int> arr, int left, int mid, int right) {
    final temp = <int>[];
    int i = left, j = mid + 1;

    while (i <= mid && j <= right) {
      temp.add(arr[i] <= arr[j] ? arr[i++] : arr[j++]);
    }
    while (i <= mid) temp.add(arr[i++]);
    while (j <= right) temp.add(arr[j++]);

    for (int k = 0; k < temp.length; k++) arr[left + k] = temp[k];
  }
}

void main() {
  final arr = [64, 25, 12, 22, 11];
  MergeSort.sort(arr);
  print(arr);
}

File: SortVision/public/code/merge/java/mergeSort.java
public class MergeSort {
    // Time: Best/Avg/Worst O(n log n), Space: O(n)
    public static void mergeSort(int[] arr) {
        if (arr == null || arr.length <= 1) return;
        sort(arr, 0, arr.length - 1);
    }

    private static void sort(int[] arr, int left, int right) {
        if (left >= right) return;
        int mid = left + (right - left) / 2;
        sort(arr, left, mid);
        sort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }

    private static void merge(int[] arr, int left, int mid, int right) {
        int[] temp = new int[right - left + 1];
        int i = left, j = mid + 1, k = 0;

        while (i <= mid && j <= right) {
            temp[k++] = arr[i] <= arr[j] ? arr[i++] : arr[j++];
        }
        while (i <= mid) temp[k++] = arr[i++];
        while (j <= right) temp[k++] = arr[j++];

        for (int x = 0; x < temp.length; x++) arr[left + x] = temp[x];
    }

    public static void main(String[] args) {
        int[] arr = {64, 25, 12, 22, 11};
        mergeSort(arr);
        for (int x : arr) System.out.print(x + " ");
        System.out.println();
    }
}
